File: typescript/packages/node/src/vfs/disk/disk.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { BoundVFS } from '@struktoai/mirage-core/vfs/bound'
import { DISK_IO } from '../../commands/builtin/disk/io.ts'
import {
  chmod,
  mkdir,
  readFile,
  stat as fsStat,
  statfs as fsStatfs,
  writeFile,
} from 'node:fs/promises'
import path from 'node:path'

import type { RegisteredCommand } from '@struktoai/mirage-core/commands/config'
import type { RegisteredOp } from '@struktoai/mirage-core/ops/registry'

import type { FindOptions, VFS } from '@struktoai/mirage-core/vfs/base'
import { CapacityState, PathSpec, VFSName } from '@struktoai/mirage-core/types'
import type { CapacityResult } from '@struktoai/mirage-core/types'

import { DISK_COMMANDS } from '../../commands/builtin/disk/index.ts'

import { readEntries, resolveInside } from '../../core/disk/utils.ts'
import { DiskAccessor } from '../../accessor/disk.ts'
import { DISK_OPS } from '../../ops/disk/index.ts'
import { DISK_PROMPT } from './prompt.ts'
import { type DeltaHook } from '@struktoai/mirage-core/watch/index'
import { buildDeltaHook } from '../../core/disk/watch/index.ts'

export interface DiskVFSOptions {
  root: string
}

export interface DiskVFSState {
  type: string
  files: Record<string, Uint8Array>
  modes?: Record<string, number>
}

async function walkFiles(current: string, out: string[]): Promise<void> {
  const entries = await readEntries(current)
  for (const e of entries) {
    const child = path.join(current, e.name)
    if (e.isDirectory()) {
      await walkFiles(child, out)
    } else if (e.isFile()) {
      out.push(child)
    }
  }
}

export class DiskVFS extends BoundVFS<DiskAccessor> implements VFS {
  declare writeFile: (p: PathSpec, data: Uint8Array) => Promise<void>

  declare appendFile: (p: PathSpec, data: Uint8Array) => Promise<void>

  declare exists: (p: PathSpec) => Promise<boolean>

  declare mkdir: (p: PathSpec, options?: { recursive?: boolean }) => Promise<void>

  declare rmdir: (p: PathSpec) => Promise<void>

  declare unlink: (p: PathSpec) => Promise<void>

  declare rename: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare truncate: (p: PathSpec, length: number) => Promise<void>

  declare copy: (src: PathSpec, dst: PathSpec) => Promise<void>

  declare rmR: (p: PathSpec) => Promise<void>

  declare du: (p: PathSpec) => Promise<number>

  declare find: (p: PathSpec, options?: FindOptions) => Promise<string[]>

  readonly kind = VFSName.DISK
  readonly cachesReads: boolean = false
  // byte store: stat() sizes every file from metadata
  readonly sizesAlwaysKnown: boolean = true
  override readonly indexTtl: number = 60
  readonly prompt = DISK_PROMPT
  readonly root: string
  readonly accessor: DiskAccessor

  constructor(options: DiskVFSOptions) {
    super(DISK_IO)
    this.root = path.resolve(options.root)
    this.accessor = new DiskAccessor(this.root)
  }

  // The resolved root is the storage: two DiskVFS instances built on the same
  // directory are one store, however they were spelled.
  override storageId(): string {
    return `${this.kind}:${this.root}`
  }

  override async open(): Promise<void> {
    await mkdir(this.root, { recursive: true })
  }

  // A real filesystem reports real numbers (QUOTA). GNU df: used counts
  // reserved blocks (blocks - bfree), available excludes them (bavail).
  override async statfs(): Promise<CapacityResult> {
    const st = await fsStatfs(this.root)
    const bsize = st.bsize
    return {
      state: CapacityState.QUOTA,
      total: st.blocks * bsize,
      used: (st.blocks - st.bfree) * bsize,
      available: st.bavail * bsize,
      inodes: st.files,
      inodesUsed: st.files - st.ffree,
      inodesFree: st.ffree,
    }
  }

  ops(): readonly RegisteredOp[] {
    return DISK_OPS
  }

  commands(): readonly RegisteredCommand[] {
    return DISK_COMMANDS
  }

  deltaHook(): DeltaHook {
    return buildDeltaHook(this.accessor)
  }

  override async getState(): Promise<DiskVFSState> {
    await mkdir(this.root, { recursive: true })
    const files: Record<string, Uint8Array> = {}
    const modes: Record<string, number> = {}
    const fileList: string[] = []
    await walkFiles(this.root, fileList)
    for (const full of fileList) {
      const rel = path.relative(this.root, full).split(path.sep).join('/')
      const data = await readFile(full)
      files[rel] = new Uint8Array(data.buffer, data.byteOffset, data.byteLength)
      // Capture the real inode mode: it is the base truth for disk
      // permissions (the sidecar is gone), so restore must reapply it or
      // a chmod would reset to the host umask.
      modes[rel] = (await fsStat(full)).mode & 0o7777
    }
    return {
      type: this.kind,
      files,
      modes,
    }
  }

  override async loadState(state: DiskVFSState): Promise<void> {
    await mkdir(this.root, { recursive: true })
    for (const [rel, data] of Object.entries(state.files)) {
      if (path.isAbsolute(rel)) throw new Error(`snapshot path must be relative: ${rel}`)
      const full = await resolveInside(this.root, PathSpec.fromStrPath('/' + rel), rel)
      await mkdir(path.dirname(full), { recursive: true })
      await writeFile(full, data)
      const mode = state.modes?.[rel]
      if (mode !== undefined) await chmod(full, mode)
    }
  }
}

File: python/mirage/vfs/disk/disk.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import os
import stat
from pathlib import Path
from typing import Any

from mirage.accessor.disk import DiskAccessor
from mirage.commands.builtin.disk import COMMANDS as DISK_COMMANDS
from mirage.commands.builtin.disk.io import IO
from mirage.core.disk.utils import resolve_inside_sync, walk_entries
from mirage.core.disk.watch import build_delta_hook
from mirage.ops.disk import OPS as DISK_OPS
from mirage.types import CapacityResult, CapacityState, PathSpec, VFSName
from mirage.vfs.bound import BoundVFS
from mirage.vfs.disk.prompt import PROMPT
from mirage.watch.base import DeltaHook


class DiskVFS(BoundVFS):

    name: str = VFSName.DISK
    # byte store: stat() sizes every file from metadata
    SIZES_ALWAYS_KNOWN: bool = True
    accessor: DiskAccessor
    index_ttl: float = 60
    PROMPT: str = PROMPT

    def __init__(self, root: str) -> None:
        super().__init__(io=IO)
        self.root = Path(root).resolve()
        # The mount root is infrastructure, not a path component a caller
        # asked for, so it is created here rather than on demand by the
        # first write: writes must report ENOENT for a missing parent the
        # way GNU does. Mirrors TypeScript, where DiskVFS.open() does
        # the same `mkdir(root, {recursive: true})`.
        self.root.mkdir(parents=True, exist_ok=True)
        self.accessor = DiskAccessor(self.root)
        for fn in DISK_COMMANDS:
            self.register(fn)
        for ro in DISK_OPS:
            self.register_op(ro)

    def storage_id(self) -> str:
        # The resolved root is the storage: two DiskVFS instances built on the
        # same directory are one store, however they were spelled.
        return f"{self.name}:{self.root}"

    def delta_hook(self) -> DeltaHook:
        return build_delta_hook(self.accessor)

    async def statfs(self) -> CapacityResult:
        # A real filesystem reports real numbers (QUOTA). GNU df: used counts
        # reserved blocks (f_blocks - f_bfree), available excludes them
        # (f_bavail); both scaled by the fundamental block size.
        st = os.statvfs(self.root)
        frsize = st.f_frsize or st.f_bsize
        return CapacityResult(
            state=CapacityState.QUOTA,
            total=st.f_blocks * frsize,
            used=(st.f_blocks - st.f_bfree) * frsize,
            available=st.f_bavail * frsize,
            inodes=st.f_files,
            inodes_used=st.f_files - st.f_ffree,
            inodes_free=st.f_favail,
        )

    def get_state(self) -> dict[str, Any]:
        files: dict[str, bytes] = {}
        modes: dict[str, int] = {}
        for directory, _, names in walk_entries(self.root):
            for name in names:
                p = directory / name
                info = p.lstat()
                if stat.S_ISREG(info.st_mode):
                    rel = p.relative_to(self.root).as_posix()
                    files[rel] = p.read_bytes()
                    modes[rel] = info.st_mode & 0o7777
        return {
            "type": self.name,
            "files": files,
            "modes": modes,
        }

    def load_state(self, state: dict[str, Any]) -> None:
        files = state.get("files", {})
        modes = state.get("modes", {})
        for rel, data in files.items():
            if Path(rel).is_absolute():
                raise ValueError(f"snapshot path must be relative: {rel}")
            spec = PathSpec.from_str_path("/" + rel)
            target = resolve_inside_sync(self.root, spec, rel)
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(data)
            mode = modes.get(rel)
            if mode is not None:
                os.chmod(target, mode)
